File: leetcode/medium/739_daily_temperatures/daily_temperatures.py
from typing import List


class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        wait = [0 for _ in range(len(temperatures))]
        stack = []
        for i in range(len(temperatures)):
            while stack and temperatures[i] > temperatures[stack[-1]]:
                index = stack.pop()
                wait[index] = i - index
            stack.append(i)
        return wait

File: leetcode/medium/739_daily_temperatures/daily_temperatures.cpp
#include <stack>
#include <vector>

class Solution {
public:
  std::vector<int> dailyTemperatures(std::vector<int> &temperatures) {
    std::stack<std::pair<int, int>> temperature_and_day;
    std::vector<int> result(temperatures.size(), 0);
    for (int day = temperatures.size() - 1; day >= 0; day--) {
      while (!temperature_and_day.empty() &&
             temperature_and_day.top().first <= temperatures[day]) {
        temperature_and_day.pop();
      }
      result[day] = temperature_and_day.empty()
                        ? 0
                        : temperature_and_day.top().second - day;
      temperature_and_day.push({temperatures[day], day});
    }
    return result;
  }
};
